File: data/python/tests/test_patsy_ModelDesc_from_formula.py
import unittest
from function_patsy_ModelDesc_from_formula import PatsyParser

class TestPatsyParser(unittest.TestCase):

    def setUp(self):
        self.parser = PatsyParser()

    def test_simple_formula(self):
        formula = "y ~ x1 + x2"
        result = self.parser.parse_formula(formula)
        self.assertIn('"lhs_terms": 1', result)
        self.assertIn('"rhs_terms": 3', result)

    def test_formula_with_no_response(self):
        formula = "~ x1 + x2"
        result = self.parser.parse_formula(formula)
        self.assertIn('"lhs_terms": 0', result)
        self.assertIn('"rhs_terms": 3', result)

    def test_invalid_formula(self):
        formula = "y ~ x1 +"
        result = self.parser.parse_formula(formula)
        self.assertEqual(result, "Error: Failed to parse formula.")

    def test_non_string_input(self):
        formula = 12345
        result = self.parser.parse_formula(formula)
        self.assertEqual(result, "Error: Formula must be a string.")

    def test_formula_no_intercept(self):
        formula = "y ~ x1 + x2 - 1"
        result = self.parser.parse_formula(formula)
        self.assertIn('"rhs_terms": 2', result)

if __name__ == "__main__":
    unittest.main()

File: data/java/tests/test_patsy_ModelDesc_from_formula.java
package com.example;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

public class test_patsy_ModelDesc_from_formula {

    private function_patsy_ModelDesc_from_formula parser;

    @BeforeEach
    void setUp() {
        parser = new function_patsy_ModelDesc_from_formula();
    }

    @Test
    void testSimpleFormula() {
        String formula = "y ~ x1 + x2";
        String result = parser.parse_formula(formula);
        assertTrue(result.contains("\"lhs_terms\":1"));
        assertTrue(result.contains("\"rhs_terms\":3"));
    }

    @Test
    void testFormulaWithNoResponse() {
        String formula = "~ x1 + x2";
        String result = parser.parse_formula(formula);
        assertTrue(result.contains("\"lhs_terms\":0"));
        assertTrue(result.contains("\"rhs_terms\":3"));
    }

    @Test
    void testInvalidFormula() {
        String formula = "y ~ x1 +";
        String result = parser.parse_formula(formula);
        assertEquals("Error: Failed to parse formula.", result);
    }

    @Test
    void testNonStringInput() {
        String formula = null;
        String result = parser.parse_formula(formula);
        assertEquals("Error: Formula must be a string.", result);
    }

    @Test
    void testFormulaNoIntercept() {
        String formula = "y ~ x1 + x2 - 1";
        String result = parser.parse_formula(formula);
        assertTrue(result.contains("\"rhs_terms\":2"));
    }
}